Improve item detail fields

mvasilak opened the pull request:
* Doesn't show empty fields when editing an item of attachment type.
* Properly ends item details editing if no changes have occured when `Done` is pressed.
* Updates date modified for each saved item details change.
* Fixes issues that reloaded previous collection view values during editing.
* Delays item field edits processing in the action handler, instead of debouncing the cell text fields observation, so the view model is always up to date, even if the `Done` is pressed to quickly.

mvasilak commented on Zotero/Controllers/Database/Requests/EditItemFieldsDbRequest.swift:
```diff
@@ -60,6 +62,37 @@ struct EditItemFieldsDbRequest: DbRequest {
+    }
+}
+
+struct EditItemFieldsDbRequest: EditItemFieldsBaseRequest, DbRequest {
```
It's used in `PDFReaderActionHandler` where it is used in `perform(writeRequests: requests)`, that's why I differentiated them like that.

michalrentka replied:
Ok, I didn't see the `perform(writeRequests:)` in changes so I couldn't find a use for `EditItemFieldsDbRequest`. In this case we can just leave this as is for now.

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ItemDetailDataCreator.swift:
```diff
@@ -241,45 +239,30 @@ struct ItemDetailDataCreator {
-        }
-        return newFieldKeys
+    /// Returns filtered, ordered set of keys for fields that have non-empty values.
+    static func filteredFieldKeys(from fields: OrderedDictionary<String, ItemDetailState.Field>) -> OrderedSet<String> {
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/adda28d0d84f886ae46894669b496184615000e8

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -260,43 +260,43 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
+                if isEditing && !data.isAttachment {
+                    state.presentedFieldIds = data.fields.keys
+                } else {
+                    state.presentedFieldIds = ItemDetailDataCreator.filteredFieldKeys(from: data.fields)
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/c8209ea4908056a8041d18f9c2bd2a0e50022c0b

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ItemDetailDataCreator.swift:
```diff
@@ -241,45 +239,30 @@ struct ItemDetailDataCreator {
-                                                isTitle: false,
-                                                isTappable: isTappable,
-                                                additionalInfo: additionalInfo)
+            fields[key] = ItemDetailState.Field(key: key, baseField: baseField, name: name, value: value, isTitle: false, isTappable: isTappable, additionalInfo: additionalInfo)
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/c8209ea4908056a8041d18f9c2bd2a0e50022c0b

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -648,16 +648,28 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
         self.update(viewModel: viewModel) { state in
             state.snapshot = state.data
-            state.data.fieldIds = ItemDetailDataCreator.allFieldKeys(for: state.data.type, schemaController: self.schemaController)
+            if !state.data.isAttachment {
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/c8209ea4908056a8041d18f9c2bd2a0e50022c0b

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ItemDetailDataCreator.swift:
```diff
@@ -241,45 +239,30 @@ struct ItemDetailDataCreator {
             if key == FieldKeys.Item.accessDate, let date = Formatter.iso8601.date(from: value) {
-                additionalInfo = [.formattedDate: Formatter.dateAndTime.string(from: date),
-                                  .formattedEditDate: Formatter.sqlFormat.string(from: date)]
+                additionalInfo = [.formattedDate: Formatter.dateAndTime.string(from: date), .formattedEditDate: Formatter.sqlFormat.string(from: date)]
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/070cc09c81626edafda365ea1366c4428b556eb4

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/Models/ItemDetailState.swift:
```diff
@@ -248,6 +247,7 @@ struct ItemDetailState: ViewModelState {
     var data: Data
     var snapshot: Data?
     var promptSnapshot: Data?
+    var presentedFieldIds: OrderedSet<String>
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/8e465a78393a37c2f1599ae5ab034e38a25d46bf

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -260,43 +260,43 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
+        ) {
+            update(viewModel: viewModel) { state in
+                state.data = data
+                if state.snapshot != nil || isEditing {
```
Indeed, check https://github.com/zotero/zotero-ios/pull/1051/commits/18a6a8c0c3fd3cb0c6ed3224f3e9b6c288890656

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -700,15 +712,21 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
-                    var requests: [DbRequest] = [EndItemDetailEditingDbRequest(libraryId: state.library.identifier, itemKey: state.key)]
-                    if !updatedFields.isEmpty {
-                        requests.insert(EditItemFieldsDbRequest(key: state.key, libraryId: state.library.identifier, fieldValues: updatedFields, dateParser: dateParser), at: 0)
+                    let endEditingRequest = EndItemDetailEditingDbRequest(libraryId: state.library.identifier, itemKey: state.key)
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/20126a82df43ac98ce28e8e0b5a35ad3bb21589a

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -955,10 +973,19 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
-            DDLogError("ItemDetailActionHandler: can't store title - \(error)")
+        let request = EditItemFieldsDbResponseRequest(key: viewModel.state.key, libraryId: viewModel.state.library.identifier, fieldValues: [keyPair: viewModel.state.data.title], dateParser: dateParser)
+        perform(request: request, invalidateRealm: false) { [weak viewModel] result in
+            switch result {
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/20126a82df43ac98ce28e8e0b5a35ad3bb21589a

mvasilak commented on Zotero/Controllers/Database/Requests/EditItemFieldsDbRequest.swift:
```diff
@@ -60,6 +62,37 @@ struct EditItemFieldsDbRequest: DbRequest {
+            return item.dateModified
         }
+
+        return nil
```
Should we consider nothing changing an error? Maybe instead of an error, we could return a `(Bool, Date)` tuple instead, the `Bool` to signify if a change was made, and the  `Date` part always being `item.dateModified`?

mvasilak replied:
I'm ambivalent about this. In places where we perform this request, we mostly treat any errors the same. Especially when we perform multiple write requests at once. 

If we throw an error, we may need to differentiate how we handle certain error types. Is this fine-grained approach desirable? It will certainly complicate some logic that is currently simple. 

Previously `EditItemFieldsDbRequest` was so forgiving, that it didn't even throw `DbError.objectNotFound`. If that was a shortcut taken in the past, then it might be ok to complicate things to have better logging for edge cases.

michalrentka replied:
Ok since this is working already let's not change unnecessary things. If something changes in the future we can update this too.

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -955,10 +973,19 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
-        self.perform(request: request) { error in
-            guard let error else { return }
-            DDLogError("ItemDetailActionHandler: can't store title - \(error)")
+        let request = EditItemFieldsDbResponseRequest(key: viewModel.state.key, libraryId: viewModel.state.library.identifier, fieldValues: [keyPair: viewModel.state.data.title], dateParser: dateParser)
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/5bad313e77dca6d7526fa54d07e64b81f1cefd0c

mvasilak commented:
@michalrentka 

> Some comments for `ItemDetailState.Data` struct:
> 
>     * `isAttachment` in `Data` should probably a constant? "Normal" item can't become an attachment, so that flag won't ever change.

Changed in https://github.com/zotero/zotero-ios/pull/1051/commits/e7309d4fe097acf50a891aa51c3ad66eb0ae2d8b

>     * `creators` and `creatorIds` can be converted to `OrderedDictionary` too.

Changed in https://github.com/zotero/zotero-ios/pull/1051/commits/0ad0e2d594a2a6e07c7a67df23eea954b95e47f9

>     * `maxFieldTitleWidth`, `maxNonemptyFieldTitleWidth` don't see to be used anymore, can they be deleted?

Changed in https://github.com/zotero/zotero-ios/pull/1051/commits/d0bfd3c06a45d50abab2f1bb90e501a33895addc

>     * should `attributedTitle` maybe be exluded from equatability of this struct? If for example font size changes (from iOS accessibility settings), the `NSAttributedString` would differ even if title never changed? It's quite an edge case, but no need to compare attributed string when we already compare the original.

Changed in https://github.com/zotero/zotero-ios/pull/1051/commits/5bad313e77dca6d7526fa54d07e64b81f1cefd0c

> What happens, if the user starts editing, the `RItem.changesSyncPaused` flag is set to `true` and then the user crashes at some point while editing? When they open the app, that `RItem`'s sync is still disabled, right? Maybe on app launch/activation we should reset all of these flags to false?

This would apply only to newly created items, as editing existing items is fully live-updating now. In theory there should be only one such entry, but just to be safe, assuming they could be more, we could offer an alert with choices to the user:
- Go to the edit screen (available only when there is one such item)
- Delete all such pending items
- Save all such pending items

> I also noticed a UI bug. When you edit some field and quickly tap "Done", the field change is debounced and since we tap "Done" before the debounce triggers, our UI then shows old state. It's stored in database and synced properly.

This should be fixed in https://github.com/zotero/zotero-ios/pull/1051/commits/20126a82df43ac98ce28e8e0b5a35ad3bb21589a

michalrentka commented:
> > What happens, if the user starts editing, the `RItem.changesSyncPaused` flag is set to `true` and then the user crashes at some point while editing? When they open the app, that `RItem`'s sync is still disabled, right? Maybe on app launch/activation we should reset all of these flags to false?
> 
> This would apply only to newly created items, as editing existing items is fully live-updating now. In theory there should be only one such entry, but just to be safe, assuming they could be more, we could offer an alert with choices to the user:
> 
> * Go to the edit screen (available only when there is one such item)
> * Delete all such pending items
> * Save all such pending items

Right, I didn't notice that `changesSyncPaused` is only used when creating a new item. In that case `EndItemDetailEditingDbRequest` could probably be renamed to something like `EndItemCreationDbRequest` and call it only when we're actually create an item in `ItemDetailActionHandler.endEditing`. 

The whole `ItemDetailActionHandler.endEditing` can be changed a lot. We don't even need to compare snapshot to data at this point, everything has been stored already. Just always call the thing which is in `guard else` now and then call `EndItemDetailEditingDbRequest` for item creation so that syncing is allowed and nothing else is needed. The whole `endEditing` has leftovers from when we had Cancel + Save buttons, but it's unnecessary now. Am I misunderstanding something?

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -942,80 +880,67 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
+        field.isTappable = ItemDetailDataCreator.isTappable(key: field.key, value: field.value, urlDetector: urlDetector, doiDetector: FieldKeys.Item.isDoi)
+
+        // If a date field has it's value edited, update only additional info here, as the user may still be typing.
+        // Modify data & accessed date field values just before storing.
```
I put it there as heads up for the modification actually taking place in `delayItemFieldsEdit` function call. Is this more clear: "Date and accessed date field values may be modified when the delayed items field edit takes place."

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/1051/commits/f15d62ea8d071e6afeb0d8e61488d27c4a373e65

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailCollectionViewHandler.swift:
```diff
@@ -180,8 +188,8 @@ final class ItemDetailCollectionViewHandler: NSObject {
                 case .title:
-                    let title = viewModel.state.data.attributedTitle
-                    return collectionView.dequeueConfiguredReusableCell(using: titleRegistration, for: indexPath, item: (title, isEditing))
+                    let attributedTitle = htmlAttributedStringConverter.convert(text: viewModel.state.data.title, baseAttributes: [.font: viewModel.state.titleFont])
```
Kept in `ItemDetailState` but not as part of `Data`. Check https://github.com/zotero/zotero-ios/pull/1051/commits/06b496ddbc9a9ac5730be43050682699e60a1123

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailCollectionViewHandler.swift:
```diff
@@ -910,8 +933,8 @@ extension ItemDetailCollectionViewHandler: UICollectionViewDelegate {
-            // Tappable fields should be only tappable when not in editing mode, in case of attachment, URL is not editable, so keep it tappable even while editing.
-            guard !viewModel.state.isEditing || (viewModel.state.data.type == ItemTypes.attachment), let field = viewModel.state.data.fields[fieldId], field.isTappable else { return }
+            // Tappable fields should be only tappable when not in editing mode, or field is not editable. E.g. in case of attachment, URL is not editable, so keep it tappable even while editing.
+            guard let field = viewModel.state.data.fields[fieldId], field.isTappable, !viewModel.state.isEditing || !field.isEditable else { return }
```
That doesn't seem correct. Parent items may have fields (URL and DOI) that are both tappable and editable. It's standalone attachments that so far have those 2 fields as tappable and not editable. Is there something I'm not considering?

michalrentka replied:
You're right, I completely forgot about those, since I was thinking only about standalone attachments. Yeah let's keep this.

mvasilak commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -1045,14 +967,152 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
+            }
+            if let startTime = delayTimer.startTime, startTime + Self.delay > .now() {
+                // If deadline wasn't reached, resume timer.
+                delayTimer.resume()
```
Check https://github.com/zotero/zotero-ios/pull/1051/commits/7cda3aa26c21f2056c734bcf5419163099686bdb

michalrentka commented on Zotero/Controllers/BackgroundTimer.swift:
```diff
@@ -11,51 +11,67 @@
+        if let startTime, startTime + timeInterval <= .now(), fireIfResumedAfterInterval {
+            eventHandler?()
+        } else {
+            state = .resumed
```
What happens if `fireIfResumedAfterInterval` is false? You `resume()` the existing `timer`, which is expired and nothing happens?

mvasilak replied:
We can ommit it, then every call to `resume` after the `timeInterval` would always call the current `eventHandler`. Of course, we don't have any use as such for now, neither for the previous implementation that would not call `eventHandler`, so here is the simpler version https://github.com/zotero/zotero-ios/pull/1051/commits/edd556d08c97ade51b4b95f21ff58d74bc1a68f8

michalrentka commented on Zotero/Scenes/Detail/ItemDetail/ViewModels/ItemDetailActionHandler.swift:
```diff
@@ -648,130 +649,64 @@ struct ItemDetailActionHandler: ViewModelActionHandler, BackgroundDbProcessingAc
+                state.isSaving = isSaving
+                state.changes.insert(.editing)
+            }
+            return
```
No reason for this return now.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/1051/commits/556f0c2cdef479d278a805902072692cd7ede542

mvasilak commented:
> The whole `ItemDetailActionHandler.endEditing` can be changed a lot. We don't even need to compare snapshot to data at this point, everything has been stored already. Just always call the thing which is in `guard else` now and then call `EndItemDetailEditingDbRequest` for item creation so that syncing is allowed and nothing else is needed. The whole `endEditing` has leftovers from when we had Cancel + Save buttons, but it's unnecessary now. Am I misunderstanding something?

Indeed, check https://github.com/zotero/zotero-ios/pull/1051/commits/e183a073eb516002d587863d29f2ff827d51c6e9 & https://github.com/zotero/zotero-ios/pull/1051/commits/8e54ace17e779dd830e692fede7061b2c772d0e4
